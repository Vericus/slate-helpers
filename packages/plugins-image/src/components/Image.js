import React from "react";
import propTypes from "prop-types";
import DefaultImageRenderer from "./DefaultImageRenderer";
import validImageFormats from "../static/validImageFormats";

export const bytesToMb = bytes => `${(bytes / 1048576).toFixed(1)} MB`;

const makeCancelable = promise => {
  let hasCanceled = false;

  const wrappedPromise = new Promise((resolve, reject) => {
    promise.then(
      val => (hasCanceled ? reject(new Error("Cancelled")) : resolve(val)),
      error => (hasCanceled ? reject(new Error("Cancelled")) : reject(error))
    );
  });

  return {
    promise: wrappedPromise,
    cancel() {
      hasCanceled = true;
    }
  };
};

class Image extends React.Component {
  constructor(props) {
    super(props);
    const src = props.node.data.get("src") || "";
    this.state = { src, loading: true, error: "" };
    this.attemptBlobUpload();
  }
  componentWillUnmount() {
    if (this.uploader) this.uploader.cancel();
  }
  onImgLoad = () => {
    // Non-blobs onLoad should finish loading
    if (!this.state.src.startsWith("blob:")) {
      this.setState({ loading: false });
    }
  };

  attemptBlobUpload() {
    const { src } = this.state;
    if (src && src.startsWith("blob:")) {
      fetch(src)
        .then(resp => resp.blob())
        .then(blob => {
          const file = new File([blob], `clipboard.png`);
          this.attemptUpload(file);
        });
    }
  }

  attemptUpload(file) {
    this.setState({ loading: true });
    const { uploadImage } = this.props.options;
    if (uploadImage) {
      this.uploader = makeCancelable(uploadImage(file));
      this.uploader.promise
        .then(newUrl => {
          this.updateSrc(newUrl);
        })
        .catch(() => {
          this.updateSrc("", false, "Failed to upload file to server");
        });
    } else {
      this.setState({ loading: false });
    }
  }

  updateError = error => {
    this.setState({ error });
  };

  // Update Src should delete this block and insert a new one
  updateSrc = (src = "", record = false, error = "") => {
    this.setState({ src, loading: false, error });
    this.props.editor.change(change => {
      if (!record) change.setOperationFlag("save", false);
      change.setNodeByKey(this.props.node.key, { data: { src } });
      if (!record) change.setOperationFlag("save", true);
    });
  };

  validFileType = type => {
    if (validImageFormats.includes(type)) {
      return true;
    }
    this.updateError("Uploaded file is not an image");
    return false;
  };

  validFileSize = size => {
    const defaultMaxFileSize = 10485760;
    const maxFileSize = this.props.options.maxFileSize || defaultMaxFileSize;
    if (size <= maxFileSize) {
      return true;
    }
    this.updateError(
      `The file exceeded the maximum size of ${bytesToMb(maxFileSize)}`
    );
    return false;
  };

  handleInsertImage = event => {
    const file = event.target.files[0];
    if (
      !file ||
      !this.validFileSize(file.size) ||
      !this.validFileType(file.type)
    ) {
      return;
    }

    const newSrc = URL.createObjectURL(file);
    this.updateSrc(newSrc);

    this.attemptUpload(file);

    if (this.input) this.input.value = "";
  };

  deleteImage = e => {
    e.preventDefault();
    e.stopPropagation();

    const { node, readOnly, editor } = this.props;
    if (!readOnly) {
      editor.change(change => {
        change.removeNodeByKey(node.key);
      });
    }
  };

  selectFile = () => {
    this.input.click();
  };

  createInput = () => (
    <input
      type="file"
      accept=".gif, .jpeg, .jpg, .png, image/gif, image/jpeg, image/jpg, image/png"
      ref={ref => {
        this.input = ref;
      }}
      onChange={e => this.handleInsertImage(e)}
      hidden
    />
  );

  render() {
    const { selectFile, deleteImage, onImgLoad, updateError } = this;
    const { attributes, readOnly, editor, isSelected, options } = this.props;
    const { src, loading, error } = this.state;
    const isReadOnly = readOnly || editor.props.isReadOnly;
    const ImageRenderer = options.renderer || DefaultImageRenderer;
    return (
      <div {...attributes}>
        {!isReadOnly && this.createInput()}
        <ImageRenderer
          actions={{ selectFile, deleteImage }}
          isReadOnly={isReadOnly}
          isSelected={isSelected}
          src={src}
          loading={loading}
          error={error}
          onImgLoad={onImgLoad}
          updateError={updateError}
        />
      </div>
    );
  }
}

Image.propTypes = {
  node: propTypes.shape({
    data: propTypes.isRequired,
    key: propTypes.string.isRequired
  }).isRequired,
  options: propTypes.shape({
    uploadImage: propTypes.func,
    maxFileSize: propTypes.number
  }).isRequired,
  editor: propTypes.shape({
    change: propTypes.func.isRequired,
    props: propTypes.shape({
      isReadOnly: propTypes.bool.isRequired
    }).isRequired
  }).isRequired,
  attributes: propTypes.shape({
    "data-key": propTypes.string.isRequired
  }).isRequired,
  readOnly: propTypes.bool.isRequired,
  isSelected: propTypes.bool.isRequired
};

export default Image;
