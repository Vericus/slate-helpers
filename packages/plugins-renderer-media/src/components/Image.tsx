import * as React from "react";
import { Block } from "slate";
import { Props as GenericProps } from "../types";

interface States {
  src: string;
  uploading: boolean;
  width: string;
}

interface Props extends GenericProps {
  getSource: (block: Block) => string;
  getImageWidth: (block: Block) => string;
}

export default class Image extends React.Component<Props, States> {
  constructor(props) {
    super(props);
    this.state = {
      src: Block.isBlock(this.props.node)
        ? this.props.getSource(this.props.node)
        : "",
      width: Block.isBlock(this.props.node)
        ? this.props.getImageWidth(this.props.node)
        : "",
      uploading: false
    };
  }

  static getDerivedStateFromProps(props, state) {
    const width = props.getImageWidth(props.node);
    const src = props.getSource(props.node);
    if (width !== state.width || src !== state.src) {
      return {
        ...state,
        width,
        src
      };
    }
    return null;
  }

  render() {
    const { className, attributes, isSelected } = this.props;
    return (
      <img
        className={className}
        src={this.state.src}
        data-image-width={this.state.width}
        data-image-is-selected={isSelected}
        {...attributes}
      />
    );
  }
}
