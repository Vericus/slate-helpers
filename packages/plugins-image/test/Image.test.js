import React from "react";
import { shallow } from "enzyme";

import Image, { resetForm, bytesToMb } from "../src/components/Image";

describe("translates bytes to megabytes", () => {
  it("translates 0 to string MB", () => {
    const zero = 0;
    expect(bytesToMb(zero)).toEqual("0.0 MB");
  });
  it("translates a 5 or less digit number of MB to string MB", () => {
    const fraction = 104857;
    expect(bytesToMb(fraction)).toEqual("0.1 MB");
  });
  it("translates to 6 digit number to string MB", () => {
    const oneDigit = 1048576;
    expect(bytesToMb(oneDigit)).toEqual("1.0 MB");
  });
  it("translates to 6+ digit number to string MB", () => {
    const twoDigit = 10485760;
    expect(bytesToMb(twoDigit)).toEqual("10.0 MB");
  });
});

describe("reset input form value", () => {
  let input;
  beforeEach(() => {
    input = document.createElement("input");
  });
  it("resets string", () => {
    input.value = "first";
    expect(input.value).toEqual("first");
    resetForm(input);
    expect(input.value).toEqual("");
  });
  it("resets file", () => {
    input.value = new File([""], "new-file");
    expect(input.value).toEqual("[object File]");
    resetForm(input);
    expect(input.value).toEqual("");
  });
});

// Default props for Image
const attributes = {
  "data-key": "1"
};
const editor = {
  change: () => {},
  props: { isReadOnly: false }
};
const node = {
  data: new Map(),
  key: attributes["data-key"]
};
const options = {};

describe("<Image /> component's onImgLoad method", () => {
  let wrapper;
  beforeEach(() => {
    wrapper = shallow(
      <Image
        attributes={attributes}
        readOnly={false}
        isSelected={false}
        editor={editor}
        options={options}
        node={node}
      />
    );
  });
  it("should not change loading when src is a blob", () => {
    const initialLoadingState = true;
    wrapper.setState({
      loading: initialLoadingState,
      src: "blob:http://localhost:6006/07b3b520-7aab-4ff8-9a84-b090f3cc4351"
    });
    wrapper.instance().onImgLoad();
    expect(wrapper.state("loading")).toEqual(initialLoadingState);
  });
  it("should set loading to false when src is NOT a blob", () => {
    wrapper.setState({
      loading: true,
      src: "http://localhost:6006/07b3b520-7aab-4ff8-9a84-b090f3cc4351"
    });
    wrapper.instance().onImgLoad();
    expect(wrapper.state("loading")).toEqual(false);
  });
});

describe("attemptBlobUpload", () => {});
describe("attemptUpload", () => {});
describe("updateError", () => {
  let wrapper;
  beforeEach(() => {
    wrapper = shallow(
      <Image
        attributes={attributes}
        readOnly={false}
        isSelected={false}
        editor={editor}
        options={options}
        node={node}
      />
    );
  });
  it("updates error with text", () => {
    const errorMessage = "There was an error";
    wrapper.instance().updateError(errorMessage);
    expect(wrapper.state("error")).toEqual(errorMessage);
  });
  it("updates error with no text", () => {
    const errorMessage = "";
    wrapper.instance().updateError(errorMessage);
    expect(wrapper.state("error")).toEqual(errorMessage);
  });
});
describe("updateSrc", () => {
  it("updates src state", () => {});
  it("sets loading state to false", () => {});
  it("sets error state to empty string", () => {});
  it("updates node data to src", () => {});
});
describe("invalidImageFile", () => {
  it("accepts png", () => {});
  it("accepts jpeg", () => {});
  it("accepts jpg", () => {});
  it("accepts gif", () => {});
  it("rejects txt", () => {});
  it("accepts docx", () => {});
});
describe("exceedsMaxFileSize", () => {
  it("rejects over default", () => {});
  it("accepts default", () => {});
  it("accepts under default", () => {});
  it("rejects over custom size", () => {});
  it("accepts custom size", () => {});
  it("accepts under custom size", () => {});
});
describe("handleInsertImage", () => {
  it("updates src if valid file selected", () => {});
});
describe("deleteImage", () => {
  it("removes the component", () => {});
});
describe("selectFile", () => {
  it("clicks the correct input", () => {});
});
describe("createInput", () => {
  it("returns an input", () => {});
});

describe("<Image /> component", () => {
  it("should render <Image /> component correctly", () => {
    const wrapper = shallow(
      <Image
        attributes={attributes}
        readOnly={false}
        isSelected={false}
        editor={editor}
        options={options}
        node={node}
      />
    );
    expect(wrapper).toMatchSnapshot();
  });
});

// Expected behaviours
describe("Loading behaviour", () => {
  it("is freshly mounted blob, upload enabled. Should immediately be loading", () => {});
  it("is freshly mounted blob, upload disabled. Should immediately be not loading (loaded)", () => {});
  it("is freshly mounted static url. Should immediately be loading", () => {});
  it("is existing mount, added blob, upload enabled. Should immediately be loading", () => {});
  it("is existing mount, added blob, upload disabled. Should immediately be not loading (loaded)", () => {});
});
