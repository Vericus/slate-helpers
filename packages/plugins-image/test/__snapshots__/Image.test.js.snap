// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Image /> component Provides functionality for the img artifact 1`] = `
<div
  data-key="1"
>
  <input
    accept=".gif, .jpeg, .jpg, .png, image/gif, image/jpeg, image/jpg, image/png"
    hidden={true}
    onChange={[Function]}
    type="file"
  />
  <DefaultImageRenderer
    actions={
      Object {
        "deleteImage": [Function],
        "selectFile": [Function],
      }
    }
    error=""
    isReadOnly={false}
    isSelected={false}
    loading={true}
    onImgLoad={[Function]}
    src=""
    updateError={[Function]}
  />
</div>
`;

exports[`<Image /> component should render <Image /> component correctly 1`] = `
<div
  data-key="1"
>
  <input
    accept=".gif, .jpeg, .jpg, .png, image/gif, image/jpeg, image/jpg, image/png"
    hidden={true}
    onChange={[Function]}
    type="file"
  />
  <DefaultImageRenderer
    actions={
      Object {
        "deleteImage": [Function],
        "selectFile": [Function],
      }
    }
    error=""
    isReadOnly={false}
    isSelected={false}
    loading={true}
    onImgLoad={[Function]}
    src=""
    updateError={[Function]}
  />
</div>
`;
